File: web/src/engine/mock/explain.ts
/**
 * `explain()` for the mock engine: the plain-language "why" behind one hazard, bucket, item,
 * requirement or warning, with the arithmetic for the expert view and the sources used.
 */
import type { Citation, Explanation, ExplainRequest } from '../types';
import { chanceWithin, dayPhrase, naturalFrequency, percent, targetDays, targetMonths, usd } from '../../lib/format';
import { returnPeriodHelp } from '../../lib/labels';
import { citation } from './citations';
import { catalogueItem } from './items';
import type { ModelResult } from './model';

const DIAL: Record<string, { label: string; n: number }> = {
  one_in_10: { label: 'Common disruptions (1-in-10)', n: 10 },
  one_in_50: { label: 'Serious (1-in-50)', n: 50 },
  one_in_100: { label: 'Very serious (1-in-100)', n: 100 },
  one_in_500: { label: 'Rare catastrophes (1-in-500)', n: 500 },
};

function cites(ids: readonly string[]): Citation[] {
  return [...new Set(ids)].map((id) => citation(id)).filter((c): c is Citation => c !== undefined);
}

/** The explanation, or undefined when the id is not in this plan. */
export function explainFrom(req: ExplainRequest, r: ModelResult): Explanation | undefined {
  const o = r.output;
  const years = req.input.dials.horizon_years;
  const dial = DIAL[req.input.dials.return_period]!;
  switch (req.kind) {
    case 'hazard': {
      const h = r.register.find((x) => x.seed.id === req.id);
      if (!h) return undefined;
      const p = h.profile;
      const plain = [p.frequency_sentence];
      const buckets = p.buckets.map((b) => o.buckets.find((x) => x.id === b)?.name.toLowerCase()).filter(Boolean);
      if (buckets.length) plain.push(`When it happens, it can mean: ${buckets.join('; ')}.`);
      if (p.confidence === 'prior') plain.push('Good data does not exist for this, so the chance is an expert estimate. It is shown as a range for that reason.');
      if (p.climate_multiplier !== 1) {
        plain.push(p.climate_multiplier > 1
          ? `Around 2050 this is projected to happen more often (about ${Math.round((p.climate_multiplier - 1) * 100)}% more). Projections, not certainties.`
          : `Around 2050 this is projected to happen less often (about ${Math.round((1 - p.climate_multiplier) * 100)}% less). Projections, not certainties.`);
      }
      if (p.display === 'rare_catastrophic') {
        plain.push('It is very unlikely, so it sits in its own box and never takes over the budget. The first days of sheltering are covered by the supplies you already plan for.');
        if (p.id === 'nuclear_attack' || p.id === 'nuclear_plant_incident') {
          plain.push(
            'If it happens: get inside, stay inside, stay tuned. Go to the middle of a sturdy building or a basement, stay there for at least 24 hours unless officials say otherwise, and listen for instructions on a radio or phone.',
          );
          if (p.id === 'nuclear_plant_incident') {
            plain.push('Potassium iodide protects only the thyroid, only matters close to a plant, and is taken only when officials say so. They hand it out in the areas that need it.');
          }
        }
        if (p.id === 'terrorism') {
          plain.push('If you are caught up in an attack: getting away is the top priority. If you cannot, hide and silence your phone. Call 911 when it is safe.');
        }
      }
      const math = [
        `Events per year for your household: ${h.seed.rate} (area) × ${(h.rate / (h.seed.rate * p.climate_multiplier)).toFixed(2)} (your home) × ${p.climate_multiplier} (climate) = ${p.rate_per_year}`,
        `Chance in ${years} years: 1 − e^(−${years} × ${p.rate_per_year}) = ${percent(chanceWithin(p.rate_per_year, years))}`,
        `Range: ${p.rate_range[0]} to ${p.rate_range[1]} events per year`,
      ];
      return { title: `Why ${p.name.toLowerCase()} is on your list`, plain, math, sources: cites(p.sources) };
    }
    case 'bucket': {
      const b = o.buckets.find((x) => x.id === req.id);
      if (!b) return undefined;
      const plain: string[] = [];
      const math: string[] = [];
      const t = b.target;
      if (t.kind === 'days') {
        plain.push(`You chose ${dial.label}. ${returnPeriodHelp(req.input.dials.return_period)}`);
        plain.push(`For "${b.name.toLowerCase()}" that comes to ${targetDays(t.value, t.low, t.high)}. The range shows how unsure the underlying numbers are.`);
        math.push(`Target: the smallest ladder value d where the yearly rate of disruptions longer than d is at most 1/${dial.n} = ${1 / dial.n}`);
        math.push('Ladder: ½, 1, 2, 3, 5, 7, 10, 14, 21, 30, 45, 60, 90, 180, 365 days');
      } else if (t.kind === 'months') {
        plain.push(`Savings for ${targetMonths(t.value, t.low, t.high)} of expenses would cover the income gap you should plan for at ${dial.label}.`);
        plain.push('This is a savings goal on its own track. It is never bought out of the supplies budget.');
      } else if (t.kind === 'evacuate') {
        plain.push(`${naturalFrequency(t.p_need_10yr)} households like yours have to leave home quickly at least once in ten years.`);
      } else {
        plain.push(`${naturalFrequency(t.p_need_10yr)} households like yours need this at least once in ten years. You have ${t.done} of ${t.of} ready.`);
      }
      if (b.contributions.length) {
        plain.push(`Most of it comes from: ${b.contributions.map((c) => `${o.register.find((h) => h.id === c.hazard)?.name.toLowerCase() ?? c.hazard} (${Math.round(c.share * 100)} in 100)`).join(', ')}.`);
      }
      if (b.relief) {
        plain.push(`For an event this size, outside help plausibly arrives in about ${dayPhrase(b.relief.help_arrives_days)}, and service is mostly back in about ${dayPhrase(b.relief.mostly_restored_days)}.`);
      }
      return { title: `How we got the target for "${b.name}"`, plain, math, sources: cites([...b.sources, ...(b.relief?.sources ?? [])]) };
    }
    case 'item': {
      const item = catalogueItem(req.id);
      const planned = o.plan.months.flatMap((m) => m.items).find((i) => i.item_id === req.id);
      if (!item || !planned) return undefined;
      const plain = [planned.why, item.spec];
      if (item.look_for.length) plain.push(`Look for: ${item.look_for.join('; ')}.`);
      if (item.avoid.length) plain.push(`Avoid: ${item.avoid.join('; ')}.`);
      const math = item.free
        ? ['Free: no cost to the budget.']
        : [
            `Quantity for your household: ${planned.quantity} ${planned.unit}`,
            `Cost: about ${usd(planned.est_cost_usd)}, from a price band of ${usd(item.price_band_usd.low)}–${usd(item.price_band_usd.high)} per ${item.price_band_usd.per}`,
          ];
      return { title: `Why "${item.name}" is in your plan`, plain, math, sources: cites(item.citations) };
    }
    case 'requirement': {
      const line = o.requirements.find((x) => x.id === req.id);
      if (!line) return undefined;
      return {
        title: `How much: ${line.plain.replace(/\.$/, '')}`,
        plain: [line.plain],
        math: [`Rule ${line.rule}: ${line.quantity} ${line.unit} for the whole household (per ${line.per})`],
        sources: cites(line.citations),
      };
    }
    case 'warning': {
      const w = o.warnings.find((x) => x.id === req.id);
      if (!w) return undefined;
      return { title: w.message, plain: [w.why, 'This is a warning, not a rule. Keep your plan as it is if you know something the model does not.'], sources: [] };
    }
  }
}

File: crates/rr-plan/src/explain.rs
//! `explain`: why a number is what it is, for a hazard, bucket, catalogue item, requirement line
//! or warning. Plain sentences first, then the arithmetic (the household event rate, Λ at the
//! target, the value integral, the quantity formula), then the sources.

use rr_content::Content;
use rr_types::{
    BucketId, BucketKind, CitationId, EngineError, ExplainKind, Explanation, HazardId, Problem,
    ProblemCode, Target,
};

use crate::coverage::Excluded;
use crate::packet::text;
use crate::pipeline::Assessment;
use crate::provenance;

fn not_found(kind: ExplainKind, id: &str) -> EngineError {
    EngineError::bad_input(vec![Problem::new(
        ProblemCode::Schema,
        "id",
        format!("There is no {kind} with the id \"{id}\" in this plan. Pick one from the plan."),
    )])
}

fn sources(ids: &[CitationId], content: &Content) -> Vec<rr_types::Citation> {
    let mut v: Vec<CitationId> = Vec::new();
    for id in ids {
        if !v.contains(id) {
            v.push(id.clone());
        }
    }
    provenance::resolve(&v, content).0
}

fn chance(rate: f64, years: f64) -> f64 {
    if rate <= 0.0 {
        0.0
    } else {
        -rr_types::math::exp_m1(-rate * years)
    }
}

/// Explains one thing in an assessed plan.
pub fn explain(
    a: &Assessment,
    content: &Content,
    kind: ExplainKind,
    id: &str,
) -> Result<Explanation, EngineError> {
    match kind {
        ExplainKind::Hazard => hazard(a, content, id),
        ExplainKind::Bucket => bucket(a, content, id),
        ExplainKind::Item => item(a, content, id),
        ExplainKind::Requirement => requirement(a, content, id),
        ExplainKind::Warning => warning(a, content, id),
    }
}

fn hazard(a: &Assessment, content: &Content, id: &str) -> Result<Explanation, EngineError> {
    let h: HazardId = id.parse().map_err(|_| not_found(ExplainKind::Hazard, id))?;
    let years = f64::from(a.input.dials.horizon_years.max(1));
    let Some(p) = a.hazards.profiles.iter().find(|p| p.id == h) else {
        return Ok(Explanation {
            title: h.name().to_owned(),
            plain: vec![format!(
                "{} is too rare where you live to list (under 1 in 100,000 a year), or does not \
                 apply to your household.",
                h.name()
            )],
            math: None,
            sources: Vec::new(),
        });
    };
    let mut plain = vec![
        p.frequency_sentence.clone(),
        rr_hazards::why_we_think_this(h).to_owned(),
    ];
    if let Some(s) = a.consequence.scenarios.iter().find(|s| {
        a.hazards
            .scenarios
            .iter()
            .any(|c| c.id == s.id && c.hazard == h)
    }) {
        plain.push(format!("{} {}", s.applies_because, s.effect_summary));
    }
    let mut math = vec![
        format!(
            "Household event rate r = λ · a · m = {:.4} a year (plausible range {:.4} to {:.4}).",
            p.rate_per_year, p.rate_range[0], p.rate_range[1]
        ),
        format!(
            "Chance of at least one in a year: 1 − e^(−r) = {:.4}.",
            p.annual_probability
        ),
        format!(
            "Of 100 households over {years} years: 100 · (1 − e^(−{years} · r)) = {:.1}.",
            100.0 * chance(p.rate_per_year, years)
        ),
        format!(
            "Severity on the fixed $50 to $500,000 scale: {:.2}.",
            p.severity
        ),
    ];
    if (p.climate_multiplier - 1.0).abs() > 1e-9 {
        math.push(format!(
            "Climate around 2050 multiplies the rate by {:.2}.",
            p.climate_multiplier
        ));
    }
    Ok(Explanation {
        title: p.name.clone(),
        plain,
        math: Some(math),
        sources: sources(&p.sources, content),
    })
}

fn bucket(a: &Assessment, content: &Content, id: &str) -> Result<Explanation, EngineError> {
    let b: BucketId = id.parse().map_err(|_| not_found(ExplainKind::Bucket, id))?;
    let ba = a.bucket(b);
    let mut plain = ba.frequency_sentences.clone();
    let mut math: Vec<String> = Vec::new();
    let rate = a.consequence.dial_rate;
    let weight = rr_budget::weights::harm_weight(b, &a.input).0;
    match ba.target {
        Target::Days { value, low, high } => {
            plain.insert(
                0,
                format!(
                    "Be ready for {} ({} is enough for this need).",
                    text::target_days(f64::from(value), f64::from(low), f64::from(high)),
                    text::lower_first(ba.tier_enough.name())
                ),
            );
            if let Some(d) = a.consequence.details.iter().find(|d| d.bucket == b) {
                let t = f64::from(value);
                math.push(format!(
                    "Λ(d) = Σ r_h · q_h,b · S_h,b(d): disruptions a year lasting longer than d \
                     days. Dial rate Λ* = {rate:.5} a year (about 1 in {}).",
                    a.input.dials.return_period.years()
                ));
                math.push(format!(
                    "Design duration: the smallest d with Λ(d) ≤ Λ* is {:.2} days; on the day \
                     ladder {} days (10th to 90th percentile {} to {}).",
                    d.target_days, value, low, high
                ));
                if t > 0.0 {
                    math.push(format!(
                        "Λ at the target: Λ({value}) = {:.5} a year.",
                        d.curve.lambda(t)
                    ));
                    let v = d.curve.value_between(0.0, t);
                    math.push(format!(
                        "Value of holding days 0 to {value}: ∫₀^{value} Λ(t) dt = {v:.4} \
                         disruption-days a year; over ten years at harm weight {weight}: 10 × \
                         {weight} × {v:.4} = {:.3}.",
                        10.0 * weight * v
                    ));
                }
                math.push(format!(
                    "Consumption: Σ r · q · E[D] = {:.3} days a year.",
                    d.consumption_days_per_year
                ));
                let table: Vec<String> = d
                    .dial_table
                    .iter()
                    .map(|p| format!("1 in {}: {} days", p.return_period_years, p.ladder_days))
                    .collect();
                math.push(format!("Targets at each setting: {}.", table.join("; ")));
            }
            let parts = a.offers.rule.parts_of(b);
            if !parts.is_empty() {
                let covered = crate::coverage::target_days(&ba.covered);
                let names: Vec<String> = parts.iter().map(|p| p.name.clone()).collect();
                math.push(format!(
                    "Covered by the plan: {covered} of {value} days, the weakest of its parts \
                     ({}).",
                    names.join(", ")
                ));
            }
        }
        Target::Months { value, low, high } => {
            plain.insert(
                0,
                format!(
                    "Aim for {} of expenses in savings.",
                    text::target_months(f64::from(value), f64::from(low), f64::from(high))
                ),
            );
            let d = &a.consequence.income;
            math.push(format!(
                "Months of income gap at the dial: {:.2}; on the months ladder {}.",
                d.target_months, d.ladder_months
            ));
            let table: Vec<String> = d
                .dial_table
                .iter()
                .map(|(y, _, l)| format!("1 in {y}: {l} months"))
                .collect();
            math.push(format!("Targets at each setting: {}.", table.join("; ")));
        }
        Target::Evacuate {
            p_need_10yr,
            notice_hours_low,
            notice_hours_high,
            days_away,
        } => {
            math.push(format!(
                "Ten-year chance of having to leave: P = 1 − e^(−10 · Σ r · q) = {p_need_10yr:.3} \
                 ({:.4} events a year).",
                a.consequence.evacuate.rate_per_year
            ));
            math.push(format!(
                "Warning: {} to {} hours; typical days away: {days_away}.",
                notice_hours_low, notice_hours_high
            ));
        }
        Target::Readiness {
            p_need_10yr,
            done,
            of,
        } => {
            math.push(format!(
                "Ten-year chance of needing it: P = 1 − e^(−10 · Σ r · q) = {p_need_10yr:.3}."
            ));
            if b.kind() == BucketKind::Readiness && of > 0 {
                plain.push(format!(
                    "The plan's checklist for it: {done} of {of} steps."
                ));
            }
        }
    }
    let mut ids = ba.sources.clone();
    if let Some(r) = &ba.relief {
        plain.push(format!(
            "For the kind of disruption behind this target, outside help plausibly arrives in \
             about {} and service is mostly back in about {}.",
            text::day_phrase(f64::from(r.help_arrives_days).round().max(0.5)),
            text::day_phrase(f64::from(r.mostly_restored_days).round().max(0.5))
        ));
        ids.extend(r.sources.iter().cloned());
    }
    if weight > 0.0 {
        ids.push(CitationId::from(rr_budget::weights::HARM_WEIGHT_CITATION));
    }
    Ok(Explanation {
        title: ba.name.clone(),
        plain,
        math: (!math.is_empty()).then_some(math),
        sources: sources(&ids, content),
    })
}

fn item(a: &Assessment, content: &Content, id: &str) -> Result<Explanation, EngineError> {
    let it = content
        .item(id)
        .ok_or_else(|| not_found(ExplainKind::Item, id))?;
    let mut plain = vec![it.spec.clone()];
    let mut math: Vec<String> = Vec::new();
    let mut ids = it.citations.clone();
    match a.offers.get(id) {
        Some(o) => {
            plain.push(format!(
                "Your household needs {}.",
                text::quantity(o.quantity, &it.unit).trim()
            ));
            if let Some((m, p)) = a
                .budget
                .plan
                .months
                .iter()
                .flat_map(|m| m.items.iter().map(move |i| (m.index, i)))
                .find(|(_, i)| i.item_id == id && i.kind != rr_types::PlanItemKind::Reserve)
            {
                plain.push(format!("In the plan in month {m}. {}", p.why));
                math.push(format!(
                    "Risk reduced (expected weighted disruption-days covered per decade): {:.3}.",
                    p.risk_reduction
                ));
            } else {
                plain.push(
                    "The plan does not buy it: other items meet the same need for less, or the \
                     need is already covered."
                        .to_owned(),
                );
            }
            for j in &o.joins {
                if let Some(l) = a.lines.iter().find(|l| l.line.id == j.line_id) {
                    if j.via_alternative && j.bucket.kind() != rr_types::BucketKind::Duration {
                        // A staging step: packed in the bag from household supplies.
                        math.push(format!(
                            "Counts toward {}: packed in it from household supplies, as part of \
                             that step rather than instead of it.",
                            l.line.id
                        ));
                    } else {
                        math.push(format!(
                            "Meets {}: needs {} {}; each {} counts as {:.3} {}.",
                            l.line.id,
                            text::number(l.quantity),
                            l.line.unit,
                            it.unit,
                            j.units_per_item,
                            l.line.unit
                        ));
                    }
                    math.extend(l.math.iter().cloned());
                    ids.extend(l.line.citations.iter().cloned());
                }
            }
        }
        None => {
            let why = a
                .offers
                .excluded
                .iter()
                .find(|(x, _)| x == id)
                .map(|(_, r)| *r);
            plain.push(
                match why {
                    Some(Excluded::NotNeeded) => {
                        "Your household does not need it (its sizing rule gives zero for you)."
                    }
                    Some(Excluded::OptionalOnly) => {
                        "It is an optional upgrade, not needed to meet your targets."
                    }
                    Some(Excluded::HazardTooRare) => {
                        "It helps with a hazard that is rare where you live, so the plan leaves \
                         it out."
                    }
                    Some(Excluded::UnknownRule) | None => "It is not part of your plan.",
                }
                .to_owned(),
            );
        }
    }
    plain.push(format!(
        "Typical price: {} per {}.",
        text::band(
            f64::from(it.price_band_usd.low),
            f64::from(it.price_band_usd.high)
        ),
        it.price_band_usd.per
    ));
    Ok(Explanation {
        title: it.name.clone(),
        plain,
        math: (!math.is_empty()).then_some(math),
        sources: sources(&ids, content),
    })
}

fn requirement(a: &Assessment, content: &Content, id: &str) -> Result<Explanation, EngineError> {
    let l = a
        .lines
        .iter()
        .find(|l| l.line.id == id)
        .ok_or_else(|| not_found(ExplainKind::Requirement, id))?;
    let mut plain = vec![l.line.plain.clone()];
    if l.prior {
        plain.push("Some amounts in it are expert estimates.".to_owned());
    }
    let mut math = l.math.clone();
    math.push(format!(
        "{} {} for the household ({}), {} step.",
        text::number(l.quantity),
        l.line.unit,
        l.line.per,
        text::lower_first(l.tier.name())
    ));
    Ok(Explanation {
        title: format!(
            "{}: {}",
            l.line.bucket.name(),
            l.line.rule.replace('_', " ")
        ),
        plain,
        math: Some(math),
        sources: sources(&l.line.citations, content),
    })
}

fn warning(a: &Assessment, content: &Content, id: &str) -> Result<Explanation, EngineError> {
    let w = a
        .warnings
        .iter()
        .find(|w| w.id == id)
        .ok_or_else(|| not_found(ExplainKind::Warning, id))?;
    let mut ids: Vec<CitationId> = Vec::new();
    for r in &w.related {
        if let Ok(b) = r.parse::<BucketId>() {
            ids.extend(a.bucket(b).sources.iter().cloned());
        }
        if let Some(s) = a.consequence.scenarios.iter().find(|s| &s.id == r) {
            ids.extend(s.sources.iter().cloned());
        }
    }
    Ok(Explanation {
        title: w.message.clone(),
        plain: vec![w.why.clone()],
        math: None,
        sources: sources(&ids, content),
    })
}
